kobotoolbox/kpi
# feat(projectHistoryLogs): record project history logs for imports TASK-944

### 🗒️ Checklist

1. [x] run linter locally
2. [x] update all related docs (API, README, inline, etc.), if any
3. [x] draft PR with a title `<type>(<scope>)<!>: <title> TASK-1234`
4. [x] tag PR: at least `frontend` or `backend` unless it's global
5. [x] fill in the template below and delete template comments
6. [x] review thyself: read the diff and repro the preview as written
7. [x] open PR & confirm that CI passes
8. [x] request reviewers, if needed
9. [ ] delete this section before merging

### 📣 Summary
Record every time a user replaces the form on a survey via an import

### 👀 Preview steps
1. ℹ️ Create a survey called MyProject
2. Go to `Form` -> `...` -> `Download XLS` and save the file as 'MyProject.xls'
3. Download 
[XLSForm_Settings.xlsx](https://github.com/user-attachments/files/17650549/XLSForm_Settings.xlsx)
4. Go to `Form` and click the 'Replace Form' icon 
5. Select `Upload an XLS form` and upload MyProject.xls
7. Go to `api/v2/audit-logs/?q=metadata__asset_uid:<asset uid> AND log_type='project-history'`
8. 🟢 There should a log with the action `replace-form` (order doesn't matter) and the usual metadata plus `latest_version_uid`.
9. Click `Replace Form` again and upload XLSForm_Settings.xlsx. This should change the title of the project.
10. Reload the endpoint
11.  🟢  Reload the endpoint. There should be two new logs, one with action=`update-name` and the other with action=`replace-form` (order doesn't matter). The log metadata should have all the information as the previous one. The `update-name` log should additionally have 
```
'name': {
   'old': <oldname>,
   'new': 'Sample Settings Form',
}
```
12. Upload MyProject.xls to Google Drive and open in Google Sheets
13. Go to File -> Share -> Publish to Web and publish as xlsx (entire document). Copy the link.
14. Back in KoboToolbox, replace the from again and select `Import an XLS from via URL`
15. Paste the Google Drive url and import
16. 🟢  Reload the endpoint. There should be a new `replace-form` log and a new `update-name` log.


### 💭 Notes
Several things in this PR
1. Pass ip address and source details from ImportTaskView to the task
2. Update `ImportTask._run_task` to record details about the asset being updated in the `messages` field
3. Update `import_in_background` to return the uid of the task, which will be sent as `result` in the `task_success` signal
4. Add a signal receiver for `task_success` that fetches the ImportTask from the uid and calls a new method in ProjectHistoryLog to create a log (or several) from it
5. Add a new method in ProjectHistoryLog that creates a `replace-form` log and possibly an `update-name` log if the name was changed by the import
6. Extra change: add `log_subtype` to the fields validated in `_check_common_metadata`
7. Extra change: change the questions in the `asset_with_settings_and_qa` fixture to make it easier to distinguish from other assets

The unit tests are a little redundant, but they are meant to check slightly different things. The ones in `test_project_history_logs` test the whole flow. The ones in `test_models` test that we know how to make logs from import tasks after they run. The ones in `test_api_imports` check that the import tasks are storing the correct information.
